File: Leetcode/python/002_Add_Two_Numbers.py
"""
 * @FileName : 002_Add_Two_Numbers.py
 * @Project : LeetCode
 * @Date : 2020-08-28
 * @author : AoN
 * @Link : https://leetcode.com/problems/add-two-numbers/
 * @Description :
 * 
"""
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        dummyHead = ListNode()
        cur = dummyHead
        carry = 0

        while l1 or l2 or carry:
            val1 = l1.val if l1 else 0
            val2 = l2.val if l2 else 0
            carry, root = divmod(val1 + val2 + carry, 10)

            cur.next = ListNode(root)
            cur = cur.next

            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None

        return dummyHead.next

File: Leetcode/java/002_Add_Two_Numbers.java
/**
 * @FileName : Solution.java
 * @Project : ALGO
 * @Date : 2020. 8. 28
 * @author : "AoN"
 * 
 * @Description : Elementary Math
 * @Link : https://leetcode.com/problems/add-two-numbers/
 * 
 */

/**
 * input
 * [2,4,3]
 * [5,6,4]
 *
 * ouput
 * [7,0,8]
 */

public class Solution {

	public class ListNode {
		int val;
		ListNode next;
		
		ListNode() {}
		ListNode(int val) {	this.val = val; }
		ListNode(int val, ListNode next) { this.val = val; this.next = next; }
	}

	public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
		ListNode dummyHead = new ListNode(0);
		ListNode p = l1, q = l2;
		ListNode cur = dummyHead;
		
		int carry = 0;
	    while (p != null || q != null) {
	    	int x, y;
	    	if(p != null) {
	    		x = p.val;
	    		p = p.next;
	    	}else {
	    		x = 0;
	    	}
	    	
	    	if(q != null) {
	    		y = q.val;
	    		q = q.next;
	    	}else {
	    		y = 0;
	    	}

	        int sum = carry + x + y;
	        carry = sum / 10;
	        cur.next = new ListNode(sum % 10);
	        cur = cur.next;
	    }
	    if (carry > 0) {
	        cur.next = new ListNode(carry);
	    }
	    return dummyHead.next;
	}
	
	
	static ListNode l1, l2;
	public static void main(String[] args) {
		Solution s = new Solution();
		
		s.addTwoNumbers(l1, l2);
	}
}
